### PR TITLE
Implement VQVAE Model for HipMRI Data

### PR DESCRIPTION
**Pull request by bairuvan**
This pull request adds a new VQVAE model for generating images from the CSIRO HipMRI Study dataset. It includes the following components:
- `modules.py`: Contains the VQVAE model implementation.
- `dataset.py`: Data loader for processing the MRI images.
- `train.py`: Training script that includes validation and testing functionalities.
- `predict.py`: Script for demonstrating the trained model on test data.
- Updated `README.md`: Documentation for setup, usage, and results.
- resources/: which contains all the graphs and images used in README.md

Cheers, Bairu An (s4702833)

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Very clear reconstructed images and best SSIM over 0.8 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

3. Design:
Modularized and good (1 Mark)

4. Commenting:
Meaningful docstrings and comments (1 Mark)

5. Algorithm above Normal Difficulty: (5 Marks)

6. Algorithm is Hard difficulty: (5 Marks)

Suggestions:

- You could also justify the hyperparameter tuning process (e.g. why you ended up choosing LR of 1e-4, what else you had attempted and what were their implications);
- Good loss/SSIM plot but could enrich the explanation on it within the Training Procedure;
- readme_images are not loaded/shown properly in the README.md due to the wrong path configuration;
- Some uploaded files are not really needed.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Pull Request has problems: Late submission.</td><td>-2</td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by bairuvan** [comment]
@shakes76 I noted that on blackboard I had lost 2 marks for late submission despite having a extension for this report. I was wondering if this was something i needed to worry about. PS i love that u have a csgo lineup playlist.

**Comment by shakes76** [comment]
Can you send me your extension outcome please?

**Comment by bairuvan** [comment]
@shakes76 i have forwarded the email containing my extension outcome.

**Comment by shakes76** [comment]
Valid extension + 2
